### ISSUE TITLE
DirectedAcyclicGraph does not throw specific exceptions for found cycles

### ISSUE THREAD

**Issue by EnderCrypt** [jgrapht/jgrapht#1099]
```
 * JGraphT version: 1.5.1 and 1.5.2-SNAPSHOT
 * Java version:    openjdk version "11.0.11" 2021-04-20
```

**Issue**
on ``DirectedAcyclicGraph``, when ``.addEdge(v,v)`` is called with vertexes that cause a cycle, an ``IllegalArgumentException`` (with message: "Edge would induce a cycle") is thrown.

the same type of exception (with message: "no such vertex in graph") happends on calling ``.addEdge(v,v)`` with vertexes that havent been added yet via ``.addVertex(v)``.

this makes it annoying to actually catch and handle the "cycle found" problem, as you cant easily determine (programatically) if the ``IllegalArgumentException`` was thrown due to it not having been added as an edge, or if it was thrown due to a cycle, ofcourse you can read it manually, but it would be significantly better to use a different exception as to allow us to ``catch(`` it and handle it in anyway we want in the program

**Steps to reproduce (small coding example)**
1. create ``new DirectedAcyclicGraph<>(DefaultEdge.class);``
2. add vertexes using ``.addVertex(v)``
3. intentionally cause a cycle using ``.addEdge(v,v)``
4. run and notice the exception thrown
5. comment out .addVertex(v) from step 2
6. run and notice the exception thrown as compared to the one on step 4

**Expected behaviour**
``.addEdge(v, v)`` should throw a custom exception upon having found a cycle, something along what already seems to exist internally ``CycleFoundException`` (perhaps make a unchecked version) which could even extend ``IllegalArgumentException`` which would preserve compatability, while allowing catching

**Comment by jsichi** [comment]
We recently addressed this for TopologicalOrderIterator in #1098, and we could do something similar here.  Let's use a separate exception since for TopologicalOrderIterator, the exception is for a traversal attempt, whereas in this case, it's for a modification attempt.  Maybe GraphCycleProhibitedException in package org.jgrapht.graph?

Would you like to submit a PR?

**Comment by EnderCrypt** [comment]
> Would you like to submit a PR?

i'll give it a shot
